### PR TITLE
CLOUD-544 Add API tests

### PR DESCRIPTION
**Pull request by blez**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: Makefile [hunk 1]
```diff
@@ -8,6 +8,9 @@ init:
 	curl -L https://github.com/uber/prototool/releases/download/v1.10.0/prototool-$(shell uname -s)-$(shell uname -m) -o ./bin/prototool
 	chmod +x ./bin/prototool
 
+	curl -L  https://github.com/go-swagger/go-swagger/releases/download/v0.25.0/swagger_$(shell uname | tr '[:upper:]' '[:lower:]')_amd64 -o ./bin/swagger
```
Comment (AlekSi) [inline Makefile:11]: It might be easier to install with Go modules

File: api-tests/apply_route_test.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	"testing"
+	"time"
+
+	"github.com/Percona-Lab/percona-version-service/client"
+	"github.com/Percona-Lab/percona-version-service/client/models"
+	"github.com/Percona-Lab/percona-version-service/client/version_service"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_apply_should_return_just_one_version(t *testing.T) {
```
Comment (AlekSi) [inline api-tests/apply_route_test.go:15]: Go doesn't use camel_case in function names

**Comment by AlekSi** [comment]
Sorry, but I can't be the first, the last, and the only reviewer. Someone from your team should review it first. Please ask for my review if you need some expertise from me. _Anyone_ can point out that snake_case is not used in Go.

**Comment by blez** [comment]
> Sorry, but I can't be the first, the last, and the only reviewer. Someone from your team should review it first. Please ask for my review if you need some expertise from me. _Anyone_ can point out that snake_case is not used in Go.

I did not add anyone for review

File: api-tests/apply_route_test.go (hunk 2 above)
Comment (blez) [reply api-tests/apply_route_test.go]: https://github.com/Percona-Lab/percona-version-service/pull/12/commits/0c336ea232130725515cf4a0c0516190fbe5e510

File: Makefile (hunk 1 above)
Comment (blez) [reply Makefile:11]: As long as installation it's already there I vote to leave it

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Review by qJkee — CHANGES_REQUESTED** [review]
As we discussed before, we do not store generated files from probuf defenitions in our repositories

**Comment by blez** [comment]
@qJkee https://github.com/Percona-Lab/percona-version-service/pull/12/commits/9bc9b241389685c410b18f06db72b4049b099e64

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Review by qJkee — APPROVED** [review]

File: Dockerfile.test [hunk 3]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.14
```
Comment (AlekSi) [inline Dockerfile.test:1]: No 1.15?

File: api-tests/apply_route_test.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+	if len(pxcResp.Payload.Versions) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Pxc) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Backup) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Proxysql) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Pmm) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Haproxy) != 1 ||
+		len(pxcResp.Payload.Versions[0].Matrix.Operator) != 1 {
+		t.Error("more than one version returned")
+	}
```
Comment (AlekSi) [inline api-tests/apply_route_test.go:36]: Looks like a custom assertion helper will be useful

File: api-tests/apply_route_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	pxcParams.WithTimeout(2 * time.Second)
+
+	for _, v := range []string{"5.7", "8.0"} {
+		pxcParams.DatabaseVersion = &v
+		pxcResp, err := cli.VersionService.VersionServiceApply(pxcParams)
+		assert.NoError(t, err)
+
+		k := getVersion(pxcResp.Payload.Versions[0].Matrix.Pxc)
+		if !strings.HasPrefix(k, v) {
```
Comment (AlekSi) [inline api-tests/apply_route_test.go:73]: assert.True ?

File: api-tests/operator_route_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		params := &version_service.VersionServiceOperatorParams{
+			OperatorVersion: c.version,
+			Product:         c.product,
+		}
+		params.WithTimeout(2 * time.Second)
+
+		resp, err := cli.VersionService.VersionServiceOperator(params)
+		assert.NoError(t, err)
+
+		assert.Equal(t, 1, len(resp.Payload.Versions))
```
Comment (AlekSi) [inline api-tests/operator_route_test.go:61]: assert.Len ?

**Review by AlekSi — COMMENTED** [review]
I will not be able to do another review, so please address my comment and merge, or just merge.

File: Dockerfile.test (hunk 3 above)
Comment (blez) [reply Dockerfile.test]: https://github.com/Percona-Lab/percona-version-service/pull/12/commits/006922ae495a7a8d108e1269117d4bf2d47d5c0d

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: api-tests/apply_route_test.go (hunk 5 above)
Comment (blez) [reply api-tests/apply_route_test.go]: https://github.com/Percona-Lab/percona-version-service/pull/12/commits/ce2cc5840e3b7b9bf117cdaeea125b5ff2ae5d51

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: api-tests/operator_route_test.go (hunk 6 above)
Comment (blez) [reply api-tests/operator_route_test.go]: https://github.com/Percona-Lab/percona-version-service/pull/12/commits/8d9b1fe668d941b92df1e4487ab40df44436c4c3

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Pull request merged**